keps/sig-architecture/2019-07-16-node-role-label-use.md

```diff
@@ -0,0 +1,215 @@
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://github.com/kubernetes/enhancements/issues
+[kubernetes/kubernetes]: https://github.com/kubernetes/kubernetes
+[kubernetes/website]: https://github.com/kubernetes/website
+
+## Summary
+
+Clarify that the `node-role.kubernetes.io/*` label is for use only by users and external projects and may not be used to vary Kubernetes behavior. Define migration process for all internal consumers of these labels.
+
+## Motivation
+
+The `node-role.kubernetes.io/master` (and the broader `node-role.kubernetes.io` namespace for labels) was introduced to provide a simple organizational and grouping convention for cluster users. The labels were reserved solely for organizing nodes via a convention that tools could recognize to display information to end users, and for use by opinionated external tooling that wished to simplify conventions for end users. Use of the label by components within the Kubernetes project (those projects subject to API review) was expressely forbidden. Specifically, no project could mandate the use of those labels in a conformant distribution, since we anticipated that many deployments of Kubernetes would have more nuanced control-plane topologies than simply "a control plane node".
+
+Over time, several changes to Kubernetes core and related projects were introduced that depended on the `node-role.kubernetes.io/master` label to vary their behavior in contravention to the guidance the label was approved under. This was unintentional and due to unclear reviewer guidelines that have since been more strictly enforced. Likewise, the complexity of Kubernetes deployments has increased and the simplistic mapping of control plane concepts to a node has proven to limit the ability of conformant Kubernetes distributions to self-host, as anticipated.
+
+
+### Goals
+
+This KEP:
+
+* Clarifies that the use of the `node-role.kubernetes/*` label namespace is reserved solely for end-user and external Kubernetes consumers, and:
+  * May not be used to vary behavior within Kubernetes projects that are subject to API review (kubernetes/kubernetes and all components that expose APIs under the `*.k8s.io` namespace)
+  * Is not required to be used for a cluster to be conformant
+* Describes the locations within Kubernetes that must be changed to use an alternative mechanism for behavior
+  * Suggests approaches for each
+* Describes the timeframe and migration process for Kubernetes distributions and deployments to update labels
+
+
+## Proposal
+
+### Use of `node-role.kubernetes.io/*` labels
+
+* Kubernetes components MUST NOT set or alter behavior on any label within the `node-role.kubernetes.io/*` namespace.
+* Kubernetes components (such as `kubectl`) MAY simplify the display of `node-role.kubernetes.io/*` labels to convey the node roles of a node
+* Kubernetes examples and documentation MUST NOT leverage the node-role labels for node placement
```

Comment:
i guess apart from a `node-role` label, a CP node would also need a regular label, that a nodeSelector can then target?

Reply:
That is not specified and is implementation specific.

There is no such thing as a "control plane node" defined in Kube, and I think targeting user workloads to those nodes is completely outside the scope of Kube.  Anyone who wants to do that needs to accept a label selector at a generic level.

I'll add that to the KEP, but I don't see this KEP as trying to solve that problem or any appetite to solve it, because control plane is vague and undefined.

Reply:
Adding to @thockin mentioned kubeadm: it would be nice if we can set node labels and taints as user via kubelet.
Maybe it’s already possible, I don’t remember and have no access to check (please forgive me if I am wrong)

This would remove a race condition, that would happen if we would have to change a node object to set these.